File: UINavigationExtensionDemo/Feature/Common/Controllers/HierarchyViewController.m
//
//  HierarchyViewController.m
//  UINavigationExtensionDemo
//
//  Created by Leo Lee on 2020/10/30.
//

#import "HierarchyViewController.h"
#import "RandomColorViewController.h"

#import "UIColor+RandomColor.h"

static CGFloat const ChooseJumpTableViewHeight = 44.0;

@interface HierarchyViewController () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSLayoutConstraint *widthConstraint;
@property (nonatomic, strong) NSLayoutConstraint *heightConstraint;

@property (nonatomic, strong) NSArray<__kindof UIViewController *> *chooseViewControllers;
@property (nonatomic, copy) void (^completionHandler)(__kindof UIViewController * _Nullable selectedViewController);

@end

@implementation HierarchyViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.5];
    [self.view addSubview:self.tableView];
    
    [self.tableView.centerXAnchor constraintEqualToAnchor:self.view.centerXAnchor].active = YES;
    [self.tableView.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor].active = YES;
    
    self.widthConstraint = [self.tableView.widthAnchor constraintEqualToConstant:0.0];
    self.widthConstraint.active = YES;

    self.heightConstraint = [self.tableView.heightAnchor constraintEqualToConstant:0.0];
    self.heightConstraint.active = YES;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    
    CGSize size = [UIScreen mainScreen].bounds.size;
    CGFloat height = self.chooseViewControllers.count * ChooseJumpTableViewHeight + 88;
    
    CGFloat maxHeight = MIN(height, size.height - 150.0);
    CGFloat maxWidth = size.width * 0.8;
    self.widthConstraint.constant = maxWidth;
    self.heightConstraint.constant = maxHeight;
}

- (void)dealloc {
    NSLog(@"%s", __FUNCTION__);
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [super touchesEnded:touches withEvent:event];
    
    __weak typeof(self) weakSelf = self;
    [self dismissViewControllerAnimated:NO completion:^{
        if (weakSelf.completionHandler) {
            weakSelf.completionHandler(nil);
        }
    }];
}

#pragma mark - Getter

- (UITableView *)tableView {
    if (!_tableView) {
        _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
        _tableView.translatesAutoresizingMaskIntoConstraints = NO;
        _tableView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
        _tableView.tableHeaderView = [[UIView alloc] init];
        _tableView.tableFooterView = [[UIView alloc] init];
        _tableView.delegate = self;
        _tableView.dataSource = self;
        _tableView.layer.cornerRadius = 10;
        [_tableView registerClass:[UITableViewCell class] forCellReuseIdentifier:NSStringFromClass([UITableViewCell class])];
    }
    return _tableView;
}

#pragma mark - UITableViewDelegate & UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Note: + 1 添加一行“创建一个新的 UIViewController”
    return self.chooseViewControllers.count + 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:NSStringFromClass([UITableViewCell class])];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:NSStringFromClass([UITableViewCell class])];
    }
    
    cell.backgroundColor = nil;
    if (indexPath.row < self.chooseViewControllers.count) {
        __kindof UIViewController *viewController = self.chooseViewControllers[indexPath.row];
        cell.textLabel.text = NSStringFromClass([viewController class]);
        cell.textLabel.textColor = [UIColor customTextColor];
        if ([viewController isKindOfClass:[BaseViewController class]]) {
            cell.backgroundColor = [(BaseViewController *)viewController randomColor];
        }
    } else {
        cell.textLabel.text = @"创建新的 UIViewController 实例对象";
        cell.textLabel.textColor = [UIColor customColorWithLightModeColor:^UIColor * _Nonnull{
            return  [UIColor blueColor];
        } darkModeColor:^UIColor * _Nonnull{
            return [[UIColor blueColor] colorWithAlphaComponent:0.5];
        }];
    }
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    __kindof UIViewController *viewController = [[RandomColorViewController alloc] init];
    if (indexPath.row < self.chooseViewControllers.count) {
        viewController = self.chooseViewControllers[indexPath.row];
    }
    __weak typeof(self) weakSelf = self;
    [self dismissViewControllerAnimated:NO completion:^{
        if (weakSelf.completionHandler) {
            weakSelf.completionHandler(viewController);
        }
    }];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return ChooseJumpTableViewHeight;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return @"选择需要跳转的视图控制器类型";
}

#pragma mark - Public

+ (void)showFromViewController:(__kindof UIViewController *)viewController
           withViewControllers:(NSArray<__kindof UIViewController *> *)viewControllers
             completionHandler:(void (^)(__kindof UIViewController * _Nullable selectedViewController))completionHandler {
    HierarchyViewController *vc = [[HierarchyViewController alloc] init];
    vc.chooseViewControllers = viewControllers;
    vc.completionHandler = completionHandler;
    vc.modalPresentationStyle = UIModalPresentationCustom;
    [viewController presentViewController:vc animated:NO completion:NULL];
}

@end

File: Example/NXNavigationExtensionSwiftDemo/Common/Comtrollers/HierarchyViewController.swift
//
//  HierarchyViewController.swift
//  NXNavigationExtensionSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit

class HierarchyViewController: BaseViewController {
    
    private static let tableViewCellHeight = CGFloat(44.0)
    
    private static let reuseIdentifier = String(describing: UITableViewCell.self)
    
    private lazy var tableView: UITableView = {
        let tableView = UITableView(frame: .zero, style: .plain)
        tableView.contentInsetAdjustmentBehavior = .never
        tableView.tableHeaderView = UIView()
        tableView.tableFooterView = UIView()
        tableView.delegate = self
        tableView.dataSource = self
        tableView.layer.cornerRadius = 10
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: HierarchyViewController.reuseIdentifier)
        return tableView
    }()
    
    private var completionHandler: ((UIViewController?) -> ())?
    private var chooseViewControllers: [UIViewController]?
    private var widthConstraint: NSLayoutConstraint?
    private var heightConstraint: NSLayoutConstraint?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.black.withAlphaComponent(0.5)
        view.addSubview(tableView)
        
        tableView.centerXAnchor.constraint(equalTo: view.centerXAnchor).isActive = true
        tableView.centerYAnchor.constraint(equalTo: view.centerYAnchor).isActive = true
        
        widthConstraint = tableView.widthAnchor.constraint(equalToConstant: 0.0);
        heightConstraint = tableView.heightAnchor.constraint(equalToConstant: 0.0);
        
        widthConstraint?.isActive = true
        heightConstraint?.isActive = true
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        guard let chooseViewControllers = chooseViewControllers else { return }
        
        let size = UIScreen.main.bounds.size
        let height = CGFloat(chooseViewControllers.count) * HierarchyViewController.tableViewCellHeight + 88.0
        let maxHeight = CGFloat.minimum(height, size.height - 150.0)
        let maxWidth = size.width * 0.8
            
        widthConstraint?.constant = maxWidth
        heightConstraint?.constant = maxHeight
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesEnded(touches, with: event)
        
        dismiss(animated: false) { [weak self] in
            self?.completionHandler?(nil)
        }
    }
    
    deinit {
        print(#function)
    }
    
}

extension HierarchyViewController: UITableViewDelegate, UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        // Note: + 1 添加一行用于 “创建一个新的 UIViewController” Item
        guard let count = chooseViewControllers?.count else { return 0 }
        return count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let reuseIdentifier = HierarchyViewController.reuseIdentifier
        let cell = tableView.dequeueReusableCell(withIdentifier: reuseIdentifier) ?? UITableViewCell(style: .default, reuseIdentifier: reuseIdentifier)
        cell.backgroundColor = nil

        if let chooseViewControllers = chooseViewControllers, indexPath.row < chooseViewControllers.count {
            let viewController = chooseViewControllers[indexPath.row]
            cell.textLabel?.text = String(describing: type(of: viewController))
            cell.textLabel?.textColor = .customText

            if let vc = viewController as? BaseViewController {
                cell.backgroundColor = vc.randomColor
            }
        } else {
            cell.textLabel?.text = "创建新的 UIViewController 实例对象"
            cell.textLabel?.textColor = UIColor.customDynamic(lightMode: { UIColor.blue }, darkMode: { UIColor.blue.withAlphaComponent(0.5) })
        }

        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        guard let chooseViewControllers = chooseViewControllers  else { return }
        var viewController: UIViewController?
        if indexPath.row < chooseViewControllers.count {
            viewController = chooseViewControllers[indexPath.row]
        } else {
            viewController = RandomColorViewController()
        }
        
        dismiss(animated: false) { [weak self] in
            self?.completionHandler?(viewController)
        }
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return HierarchyViewController.tableViewCellHeight
    }
    
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return "选择需要跳转的视图控制器类型"
    }
    
}

extension HierarchyViewController {
    
    static func showFromViewController(_ viewController: UIViewController, withViewControllers viewControllers: [UIViewController]?, completionHandler:((UIViewController?) -> ())?) {
        let vc = HierarchyViewController()
        vc.chooseViewControllers = viewControllers
        vc.completionHandler = completionHandler
        vc.modalPresentationStyle = .custom
        viewController.present(vc, animated: false, completion: nil)
    }
    
}
